src/interpreter.ts

```diff
@@ -1,13 +1,462 @@
+                    source,
+                );
+            }
+            return valLeft !== valRight;
+        case "&&":
+            return (
+                ensureBoolean(valLeft, locLeft) &&
+                ensureBoolean(valRight, locRight)
+            );
+        case "||":
+            return (
+                ensureBoolean(valLeft, locLeft) ||
+                ensureBoolean(valRight, locRight)
+            );
+    }
+}
+
+function interpretEscapeSequences(stringLiteral: string, source: SrcInfo) {
+    return stringLiteral.replace(
+        /\\\\|\\"|\\n|\\r|\\t|\\v|\\b|\\f|\\u{([0-9A-Fa-f]{1,6})}|\\u([0-9A-Fa-f]{4})|\\x([0-9A-Fa-f]{2})/g,
+        (match, unicodeCodePoint, unicodeEscape, hexEscape) => {
+            switch (match) {
+                case "\\\\":
+                    return "\\";
+                case '\\"':
+                    return '"';
+                case "\\n":
+                    return "\n";
+                case "\\r":
+                    return "\r";
+                case "\\t":
+                    return "\t";
+                case "\\v":
+                    return "\v";
+                case "\\b":
+                    return "\b";
+                case "\\f":
+                    return "\f";
+                default:
+                    // Handle Unicode code point escape
+                    if (unicodeCodePoint) {
+                        const codePoint = parseInt(unicodeCodePoint, 16);
+                        if (codePoint > 0x10ffff) {
+                            throwErrorConstEval(
+                                `unicode code point is outside of valid range 000000-10FFFF: ${stringLiteral}`,
+                                source,
+                            );
+                        }
+                        return String.fromCodePoint(codePoint);
+                    }
+                    // Handle Unicode escape
+                    if (unicodeEscape) {
+                        const codeUnit = parseInt(unicodeEscape, 16);
+                        return String.fromCharCode(codeUnit);
+                    }
+                    // Handle hex escape
+                    if (hexEscape) {
+                        const hexValue = parseInt(hexEscape, 16);
+                        return String.fromCharCode(hexValue);
+                    }
+                    return match;
+            }
+        },
+    );
+}
+
+class ReturnSignal extends Error {
+    private value?: Value;
+
+    constructor(value?: Value) {
+        super();
+        this.value = value;
+    }
+
+    public getValue(): Value | undefined {
+        return this.value;
+    }
+}
+
+type InterpreterConfig = {
+    // Options that tune the interpreter's behavior.
+
+    // Maximum number of iterations inside a loop before a time out is issued.
+    // This option only applies to: do...until and while loops
+    maxLoopIterations: bigint;
+};
+
+class RuntimeEnvironment {
+    private values: Map<string, Value>;
```

Comment:
Looks like you want a data structure which could probably be called something like "persistent multi map", i.e. a map from identifiers to stacks of values: the last key removal would actually keep the key but re-associate it with the previous value, i.e. values: `Map<string, Value[]>` and then you don't need the `enclosing` environment

Reply:
Or maybe not, since you are emulating the call stack frames with a linked list of environments. Please ignore the above, since it would require more bookkeeping

Reply:
Yeah, I was actually thinking how to implement your suggestion, because I would also need to keep track of which variables were modified at each stack frame so that I only drop those when the frame closes.